### ISSUE TITLE
Cardinality incorrect with typing.Optional typing.List for Pydantic dataclasses?

### ISSUE THREAD

**Issue by sdruskat** [drivendataorg/erdantic#105]
Hi 👋, and thanks for developing such a helpful project!

I'm wondering whether cardinality for optional lists is currently correctly defined as zero-or-one? I'm not very knowledgeable when it comes to ERDs, so this may be just wrong thinking on my side.

(This may be related to #90 as well.)

The following Pydantic model (two `Child` classes to avoid rendering issues) gives the ERD below, created by running `erdantic my_module.Parent -d | dot -Tsvg > parent.svg`. (`pygraphviz 1.12`)

```python
from typing import List, Optional
from pydantic.dataclasses import dataclass

@dataclass
class Child1:
    something: str

@dataclass
class Child2:
    something: str

@dataclass
class Parent:
    required_list: List[Child1]
    optional_list: Optional[List[Child2]]
```

![ERD for the parent model](https://github.com/drivendataorg/erdantic/assets/3007126/e7771fbb-de80-4450-aa67-7a24178bad57)

I think the cardinality (`0..*`) at `Child1` is correct: `required_list` is a list that can contain anything from zero to many instances of `Child1`.

However, the cardinality at `Child2` (`0..1`) is wrong - I think? Following the logic derived from `required_list` (`0..*` instances of `Child1` *can be included in* `required_list`), I would expect the same cardinality at `Child2` as well, meaning `0..*` instances of `Child2` *can be included in* `optional_list` (regardless of whether the list must exist, i.e., the field must be defined on instantiation).

I haven't had a look at the source code yet, but would assume that perhaps `Optional` fields are processed without looking at the type of the wrapped value?

Thanks, and do please correct me if I got it wrong.

**Comment by jayqi** [comment]
You are right. The current implementation of `is_many` only checks whether the outermost type is a container. 

This actually surfaces a deeper issue, which is that our representation of manyness is wrong. Currently, it's set as a property of the field on the parent class, when it's really a property of the edge. For example, one field with the type annotation `Union[Child1, List[Child2]]` should be one-to-one with `Child1` and one-to-many with `Child2`.

**Comment by jayqi** [comment]
This has been fixed in v1.0.0. This is temporarily only available as a pre-release version (`pip install erdantic --pre`). 

Here are the test cases that cover this:

https://github.com/drivendataorg/erdantic/blob/1eb44598e4a64f8234228818926ee0c9cd9f4283/tests/test_typing_utils.py#L15-L27
